packages/loopring_v3/circuit/Utils/Data.h

```diff
@@ -222,353 +340,371 @@ class Ring
 {
 public:
-    Ring ring;
-
-    ethsnarks::FieldT accountsMerkleRoot;
-
-    TradeHistoryUpdate tradeHistoryUpdate_A;
-    TradeHistoryUpdate tradeHistoryUpdate_B;
-
-    BalanceUpdate balanceUpdateS_A;
-    BalanceUpdate balanceUpdateB_A;
-    AccountUpdate accountUpdate_A;
-
-    BalanceUpdate balanceUpdateS_B;
-    BalanceUpdate balanceUpdateB_B;
-    AccountUpdate accountUpdate_B;
-
-    BalanceUpdate balanceUpdateA_P;
-    BalanceUpdate balanceUpdateB_P;
-
-    BalanceUpdate balanceUpdateA_O;
-    BalanceUpdate balanceUpdateB_O;
+    ethsnarks::FieldT owner;
+    ethsnarks::FieldT accountID;
+    ethsnarks::FieldT tokenID;
+    ethsnarks::FieldT amount;
+    ethsnarks::FieldT index;
 };
 
-static void from_json(const json& j, RingSettlement& ringSettlement)
+static void from_json(const json& j, Deposit& deposit)
 {
-    ringSettlement.ring = j.at("ring").get<Ring>();
-
-    ringSettlement.accountsMerkleRoot = ethsnarks::FieldT(j.at("accountsMerkleRoot").get<std::string>().c_str());
-
-    ringSettlement.tradeHistoryUpdate_A = j.at("tradeHistoryUpdate_A").get<TradeHistoryUpdate>();
-    ringSettlement.tradeHistoryUpdate_B = j.at("tradeHistoryUpdate_B").get<TradeHistoryUpdate>();
-
-    ringSettlement.balanceUpdateS_A = j.at("balanceUpdateS_A").get<BalanceUpdate>();
-    ringSettlement.balanceUpdateB_A = j.at("balanceUpdateB_A").get<BalanceUpdate>();
-    ringSettlement.accountUpdate_A = j.at("accountUpdate_A").get<AccountUpdate>();
-
-    ringSettlement.balanceUpdateS_B = j.at("balanceUpdateS_B").get<BalanceUpdate>();
-    ringSettlement.balanceUpdateB_B = j.at("balanceUpdateB_B").get<BalanceUpdate>();
-    ringSettlement.accountUpdate_B = j.at("accountUpdate_B").get<AccountUpdate>();
-
-    ringSettlement.balanceUpdateA_P = j.at("balanceUpdateA_P").get<BalanceUpdate>();
-    ringSettlement.balanceUpdateB_P = j.at("balanceUpdateB_P").get<BalanceUpdate>();
-
-    ringSettlement.balanceUpdateA_O = j.at("balanceUpdateA_O").get<BalanceUpdate>();
-    ringSettlement.balanceUpdateB_O = j.at("balanceUpdateB_O").get<BalanceUpdate>();
+    deposit.owner = ethsnarks::FieldT(j.at("owner").get<std::string>().c_str());
+    deposit.accountID = ethsnarks::FieldT(j.at("accountID"));
+    deposit.tokenID = ethsnarks::FieldT(j.at("tokenID"));
+    deposit.amount = ethsnarks::FieldT(j.at("amount").get<std::string>().c_str());
+    deposit.index = ethsnarks::FieldT(j.at("index").get<std::string>().c_str());
 }
 
-class RingSettlementBlock
+class Withdrawal
 {
 public:
-
-    ethsnarks::FieldT exchangeID;
-
-    ethsnarks::FieldT merkleRootBefore;
-    ethsnarks::FieldT merkleRootAfter;
-
-    ethsnarks::FieldT timestamp;
-
-    ethsnarks::FieldT protocolTakerFeeBips;
-    ethsnarks::FieldT protocolMakerFeeBips;
-
-    Signature signature;
-
-    AccountUpdate accountUpdate_P;
-
-    ethsnarks::FieldT operatorAccountID;
-    AccountUpdate accountUpdate_O;
-
-    std::vector<Loopring::RingSettlement> ringSettlements;
+    ethsnarks::FieldT accountID;
```

Comment:
Maybe I can figure out this one but now I don't know:
why `owner` is missing from `Withdrawal` but not `Deposit`?

Reply:
For a withdrawal the owner is always the owner as set in the account leaf the withdrawal is done for. Deposits can also be done to new accounts that don't have an owner set yet, so we take it as input here as well so we can set in on the account when necessary or check that they match.